libr/anal/anal.c

```diff
@@ -380,94 +378,135 @@ R_API int r_anal_archinfo(RAnal *anal, int query) {
+			eprintf ("Cant find Function at given address\n");
+			return false;
+		}
+		tmp_name =fcn ? fcn->name: fi->name;
+	}
+	if (r_anal_type_func_exist (anal, tmp_name)) {
+		fnl_name = strdup (tmp_name);
+	} else if (!(fnl_name = r_anal_type_func_guess (anal, (char *)tmp_name))) {
+		eprintf ("Cant find prototype for %s in types databse\n", tmp_name);
 		return false;
 	}
-	r_list_append (anal->noreturn, nr);
+	sdb_bool_set (anal->sdb_types, sdb_fmt (-1, "func.%s.noreturn", fnl_name), true, 0);
+	free (fnl_name);
 	return true;
 }
 
+static int noreturn_dropall(void *p, const char *k, const char *v) {
+	RAnal *anal = (RAnal *)p;
+	if (!strncmp (v, "func", strlen ("func") + 1)) {
+		char *query = sdb_fmt (-1, "func.%s.noreturn", k);
+		sdb_unset (anal->sdb_types, query, 0);
+	}
+	return 1;
+
+}
+
 R_API int r_anal_noreturn_drop(RAnal *anal, const char *expr) {
-	bool ret = false;
 	if (!strcmp (expr, "*")) {
-		if (!r_list_empty (anal->noreturn)) {
-			r_list_free (anal->noreturn);
-			anal->noreturn = r_list_newf ((RListFree)&r_anal_noreturn_free);
-			ret = true;
-		}
+			sdb_foreach (anal->sdb_types, noreturn_dropall, anal);
+			return true;
 	} else {
-		RListIter *iter, *iter2;
-		RAnalNoreturn *nr;
+		const char *fcnname = NULL;
+		char *tmp;
 		if (!strncmp (expr, "0x", 2)) {
 			ut64 n = r_num_math (NULL, expr);
-			r_list_foreach_safe (anal->noreturn, iter, iter2, nr) {
-				if (nr->addr == n) {
-					r_list_delete (anal->noreturn, iter);
-					ret = true;
-				}
+			RAnalFunction *fcn = r_anal_get_fcn_in (anal, n, -1);
+			if (!fcn) {
+				eprintf ("can't find function at 0x%"PFMT64x"\n", n);
+				return false;
 			}
+			fcnname = fcn->name;
 		} else {
-			r_list_foreach_safe (anal->noreturn, iter, iter2, nr) {
-				if (r_str_glob (nr->name, expr)) {
-					r_list_delete (anal->noreturn, iter);
-					ret = true;
-				}
-			}
+			fcnname = expr;
+		}
+		if (r_anal_type_func_exist (anal, fcnname)) {
+			char *query = sdb_fmt (-1, "func.%s.noreturn", fcnname);
+			sdb_unset (anal->sdb_types, query, 0);
+			return true;
+		} else if ((tmp = r_anal_type_func_guess (anal, (char *)fcnname))) {
+			char *query = sdb_fmt (-1, "func.%s.noreturn", tmp);
+			sdb_unset (anal->sdb_types, query, 0);
+			free (tmp);
+			return true;
+		} else {
+			eprintf ("Cant find prototype for %s in types databse", fcnname);
+			return false;
 		}
 	}
-	return ret;
 }
+static bool r_anal_noreturn_at_name (RAnal *anal, const char *name) {
+	bool ret = sdb_bool_get (anal->sdb_types, sdb_fmt (-1, "func.%s.noreturn", name), NULL);
+	if (ret) {
+		return true;
+	}
+	char *tmp = r_anal_type_func_guess (anal, (char *)name);
+	if (tmp) {
+		ret = sdb_bool_get (anal->sdb_types, sdb_fmt (-1, "func.%s.noreturn", tmp), NULL);
+		if (ret) {
+			return true;
+		}
+	}
 
+}
 R_API bool r_anal_noreturn_at(RAnal *anal, ut64 addr) {
 	RListIter *iter;
-	RAnalNoreturn *nr;
 	RAnalFunction *f = r_anal_get_fcn_at (anal, addr, 0);
 	RFlagItem *fi = anal->flb.get_at (anal->flb.f, addr);
```

Comment:
we dont need fi at this point so we can put this flb.get_at() after the if(f) {} block and it will be faster